pkg/scheduler/scheduling_node.go

```diff
@@ -27,67 +27,158 @@ import (
+    if sn.availableDirty {
+		sn.cachedAvailable = resources.Sub(sn.nodeInfo.GetAvailableResource(), sn.allocatingResource)
+		sn.availableDirty = false
 	}
+	return sn.cachedAvailable
+}
+
+// Get the resource tagged for allocation on this node.
+// These resources are part of unconfirmed allocations.
+func (sn *SchedulingNode) getAllocatingResource() *resources.Resource {
+	sn.lock.RLock()
+	defer sn.lock.RUnlock()
+
+	return sn.allocatingResource
+}
+
+// Update the number of resource proposed for allocation on this node
+func (sn *SchedulingNode) incAllocatingResource(proposed *resources.Resource) {
+	sn.lock.Lock()
+    defer sn.lock.Unlock()
+
+    sn.availableDirty = true
+	sn.allocatingResource.AddTo(proposed)
+}
+
+// Handle the allocation processing on the scheduler when the cache node is updated.
+func (sn *SchedulingNode) handleAllocationUpdate(confirmed *resources.Resource) {
+	sn.lock.Lock()
+	defer sn.lock.Unlock()
+	log.Logger().Debug("allocations in progress increased",
+		zap.String("nodeId", sn.NodeId),
+		zap.Any("confirmed", confirmed))
+
+	sn.availableDirty = true
+	sn.allocatingResource.SubFrom(confirmed)
+}
+
+// Get the number of resource tagged for preemption on this node
+func (sn *SchedulingNode) getPreemptingResource() *resources.Resource {
+	sn.lock.RLock()
+	defer sn.lock.RUnlock()
+
+	return sn.preemptingResource
+}
+
+// Update the number of resource tagged for preemption on this node
+func (sn *SchedulingNode) incPreemptingResource(preempting *resources.Resource) {
+	sn.lock.Lock()
+	defer sn.lock.Unlock()
+
+    sn.preemptingResource.AddTo(preempting)
+}
+
+func (sn *SchedulingNode) handlePreemptionUpdate(preempted *resources.Resource) {
+	sn.lock.Lock()
+	defer sn.lock.Unlock()
+	log.Logger().Debug("preempted resources released",
+		zap.String("nodeId", sn.NodeId),
+		zap.Any("preempted", preempted))
+
+	sn.preemptingResource.SubFrom(preempted)
+}
+
+// Check and update allocating resources of the scheduling node.
+// If the proposed allocation fits in the available resources, taking into account resources marked for
+// preemption if applicable, the allocating resources are updated and true is returned.
+// If the proposed allocation does not fit false is returned and no changes are made.
+func (sn *SchedulingNode) CheckAndAllocateResource(delta *resources.Resource, preemptionPhase bool) bool {
+	sn.lock.Lock()
+	defer sn.lock.Unlock()
+	available := sn.nodeInfo.GetAvailableResource()
+	newAllocating := resources.Add(delta, sn.allocatingResource)
 
-	if resources.FitIn(avail, newAllocating) {
-		m.AllocatingResource = newAllocating
+    if preemptionPhase {
+        available = resources.Add(available, sn.preemptingResource)
+    }
+    if resources.FitIn(available, newAllocating) {
+        log.Logger().Debug("allocations in progress updated",
+			zap.String("nodeId", sn.NodeId),
+			zap.Any("total unconfirmed", newAllocating))
+		sn.availableDirty = true
+		sn.allocatingResource = newAllocating
```

Comment:
Looks like this can be simplified to:

```
availableDirty = true
available := sn.getAvailableResource()
resources.FitIn(available, delta)
  sn. incAllocatingResource(delta)
```

Reply:
`available` is updated with the preempting resources during the preemption phase so we can not use that directly from the call. The delta is the _increase_ to what we have.
We cannot ignore the currently already in progress but not confirmed yet allocations which are stored `allocatingResource`.  The `newAllocating` has to fit in `available`, we cannot shortcut this.

Reply:
Fair point.